Enhancement to Confusion Matrix Output Representation for improving readability #19012

shubhamdo opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #19012 

#### What does this implement/fix? Explain your changes.
When you have multiple levels you can have difficulty reading the ndarray, associating the levels with the True and Predicted values. It is an enhancement to the output of confusion matrix function, better representing the true and predicted values for multilevel classes.

- Returns a confusion matrix in dict representation with labels as keys ('true', 'pred')
- Example:
```python
    >>> y_true = ["cat", "ant", "cat", "cat", "ant", "bird"]
    >>> y_pred = ["ant", "ant", "cat", "cat", "ant", "cat"]
    >>> confusion_matrix(y_true, y_pred, labels=["ant", "bird", "cat"], pprint=True)
    {('ant', 'ant'): 2, ('bird', 'ant'): 0, ('cat', 'ant'): 1,
     ('ant', 'bird'): 0, ('bird', 'bird'): 0, ('cat', 'bird'): 0,
     ('ant', 'cat'): 0, ('bird', 'cat'): 1, ('cat', 'cat'): 2}
```
#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

shubhamdo committed:
Dictionary Representation of Confusion Matrix

shubhamdo committed:
Confusion Matrix Flat Dict Representation using tuples as key

shubhamdo committed:
Merge branch 'master' into feature19012

shubhamdo committed:
Test Case for confusion_matrix_pprint Added

shubhamdo committed:
Documentation for Confusion Matrix Updated

shubhamdo committed:
Merge branch 'master' into feature19012

jnothman commented on sklearn/metrics/_classification.py:
```diff
@@ -249,6 +249,10 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
         conditions or all the population. If None, confusion matrix will not be
         normalized.
 
+    pprint : bool, default=False
```
let's call this `as_dict`?

jnothman commented on sklearn/metrics/_classification.py:
```diff
@@ -257,6 +261,14 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
         samples with true label being i-th class
         and predicted label being j-th class.
 
+    Or
```
This isn't valid numpydoc. The types need to be mentioned all on the first line.

shubhamdo replied:
1. Changed parameter name from pprint --> as_dict()
2. Changed the Testing Function in test_classification.py
3. _Changed the Docstring for the function, added explaination of Series usage_

Not sure about the numpydoc please review.
- Should I mention it as --> tuple[ndarry, dict['true_class','pred_class']] ?

jnothman commented on sklearn/metrics/_classification.py:
```diff
@@ -249,6 +249,10 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
 
+    pprint : bool, default=False
+        Returns a confusion matrix in dict representation with labels as keys
+        ('true', 'pred')
```
It would be worth briefly noting the usage with pandas and unstack.

jnothman reviewed (commented):
Thanks for the PR.

shubhamdo committed:
Added reviewed changes
1. Changed parameter name from pprint --> as_dict()
2. Changed the Testing Function in test_classification.py
3. Tested
4. Changed the Docstring for the function, added explaination of Series usage

shubhamdo committed:
Linting Issues Fixed

shubhamdo committed:
Merge branch 'master' into feature19012

shubhamdo committed:
Liniting fix _classification.py

shubhamdo committed:
Liniting fix _classification.py

shubhamdo committed:
Changelog updated

shubhamdo committed:
Linting Issue Fix

tansaku commented:
I guess this PR has stalled?
